addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/JsoScanner.java

```diff
@@ -0,0 +1,160 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.zap.extension.pscanrulesAlpha;
+
+import java.net.HttpCookie;
+import java.util.Arrays;
+import java.util.List;
+import net.htmlparser.jericho.Source;
+import org.parosproxy.paros.Constant;
+import org.parosproxy.paros.core.scanner.Alert;
+import org.parosproxy.paros.network.HttpHeaderField;
+import org.parosproxy.paros.network.HttpMessage;
+import org.zaproxy.zap.extension.pscan.PassiveScanThread;
+import org.zaproxy.zap.extension.pscan.PluginPassiveScanner;
+
+/** Java Serialized Objects (JSO) scanner. Detect the magic sequence and generate an alarm */
+public class JsoScanner extends PluginPassiveScanner {
+
+    /** Prefix for internationalized messages used by this rule */
+    private static final String MESSAGE_PREFIX = "pscanalpha.jso.";
+
+    private static final byte[] JSO_BYTE_MAGIC_SEQUENCE = {(byte) 0xac, (byte) 0xed, 0x00, 0x05};
+    private static final String JSO_BASE_64_MAGIC_SEQUENCE = "rO0AB";
+    private static final String JSO_URI_ENCODED_MAGIC_SEQUENCE = "%C2%AC%C3%AD%00%05";
+    private PassiveScanThread parent;
+
+    @Override
+    public void scanHttpRequestSend(HttpMessage msg, int id) {
+        checkJsoInQueryParameters(msg);
+
+        checkJsoInHeaders(msg, msg.getRequestHeader().getHeaders());
+
+        checkJsoInCookies(msg, msg.getRequestHeader().getHttpCookies());
+
+        checkJsoInBody(msg, msg.getRequestBody().getBytes(), msg.getRequestBody().toString());
+    }
+
+    @Override
+    public void scanHttpResponseReceive(HttpMessage msg, int id, Source source) {
+        checkJsoInHeaders(msg, msg.getResponseHeader().getHeaders());
+
+        checkJsoInCookies(msg, msg.getResponseHeader().getHttpCookies(null));
+
+        checkJsoInBody(msg, msg.getResponseBody().getBytes(), msg.getResponseBody().toString());
+    }
+
+    private void checkJsoInQueryParameters(HttpMessage msg) {
+        String escapedURI = msg.getRequestHeader().getURI().getEscapedQuery();
+        if (escapedURI == null) {
+            return;
+        }
+        String[] params = escapedURI.split("&");
+        for (String param : params) {
+            String[] strings = param.split("=");
+            if (strings.length <= 1) {
+                continue;
+            }
+            String value = strings[1];
+            if (hasUriEncodedMagicSequence(value) || hasJsoBase64MagicSequence(value)) {
+                raiseAlert(msg, "");
+            }
+        }
+    }
+
+    private void checkJsoInBody(HttpMessage msg, byte[] bytes, String body) {
+        byte[] startOfBody = Arrays.copyOfRange(bytes, 0, JSO_BYTE_MAGIC_SEQUENCE.length);
```

Comment:
I think you meant to change this part, your commit message even mentions it :zap:

Reply:
I extract the body as a parameter to make it usable for the request and the response.
Maybe you can help me :) can you explain me the idea you have in mind?

Reply:
Maybe I misunderstood @thc202's https://github.com/zaproxy/zap-extensions/pull/2121#discussion_r305918165

Reply:
Just spoke with him quickly in IRC, it seems I did misunderstand.